### PR TITLE
circleci: Add macOS tests to CircleCI (fixes #386)

### LINKED ISSUE #386: Official support for Windows & Mac
Given a lot of people use Windows or Mac for development, it'd be nice if `imitation` could reliably support this. We don't use any Linux/BSD-specific features. Moreover, PyTorch and even MuJoCo support Windows and Mac, so this should require minimal (if any) porting.

To make sure we don't break it in the future, it'd be good to add continuous integration testing for these platforms. CircleCI supports [Mac and Windows executors](https://circleci.com/docs/2.0/executor-intro/), so I think it's just a matter of adding versions of the current [`unit-test` job](https://github.com/HumanCompatibleAI/imitation/blob/dbd26c889c36d70bf0eef42dd44f050a7d328541/.circleci/config.yml#L127) for these platforms. The complicating factor will be that we'll need to setup the binary environment to support what we need -- probably we just want to skip MuJoCo and all the nasty graphics dependency it pulls in? Might want to run an abbreviated version of the test suite on these anyway, both because CircvleCI [pricing](https://circleci.com/product/features/resource-classes/) is a lot higher for Windows/Mac and to not slow down the CI test too much.

### PR DESCRIPTION
**Pull request by tomtseng**
This PR adds macOS tests to CI to make sure we're always maintaining macOS support. Fixes #386.

The macOS CI test job is similar to the existing Linux CI test job. Some differences:
- We use Python 3.9 due to avoid an [upstream Ray issue](https://github.com/ray-project/ray/issues/27380) that occurs if we use Python 3.8
- We need to install some macOS packages with `brew install coreutils ffmpeg gnu-getopt` since we don't have these packages preloaded in a Docker image like we do with the Linux test job. Relatedly, we do not set up MuJoCo (which will be removed from our CI tests anyway if PR #515 merges)

The only expected behavior change this PR will have on the Linux+Windows CI jobs is renaming some CI cache keys, as commented [here](https://github.com/HumanCompatibleAI/imitation/pull/511/files#r936749919) and [here](https://github.com/HumanCompatibleAI/imitation/pull/511/files#r936765843).

## Testing

- CI should pass for linux, windows, and macOS

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -190,74 +268,55 @@ jobs:
...
 
-      - store_test_results:
-          path: /tmp/test-reports
-          unit-test:
+      - run:
+          name: run tests
+          command: |
+            source ~/venv/bin/activate
+            # Xdummy-entrypoint.py: starts an X server and sets DISPLAY, then runs wrapped command.
+            ./ci/Xdummy-entrypoint.py pytest -n auto \
```
Comment (AdamGleave) [inline .circleci/config.yml:300]: Does Xdummy work on Mac? I thought the windowing system was different.

Also may not be necessary if CircleCI Mac executor isn't headless.

**Review by AdamGleave — COMMENTED** [review]
I like that you refactored things into subcommands (save/restore pytest cache), very DRY!

MuJoCo seems like a common pain point with CI. Taking a quick look through the tests I can only see a single test, `test_parallel_train_adversarial_custom_env` that definitely uses MuJoCo. Moreover, I think we could easily enough change it to not use MuJoCo by changing `env_named_config` to something from `"seals_ant"` to a classic control environment like `"pendulum"`.

I'd suggest trying to change that test to not use MuJoCo then run the tests in a non-MuJoCo environment and verify nothing new breaks (nor is skipped when it wasn't before). If so, we can rip MuJoCo out of all the CI tests to simplify things.

File: .circleci/config.yml (hunk 1 above)
Comment (tomtseng) [reply .circleci/config.yml]: I just copied this from the Linux tests. The `Xdummy-entrypoint.py` script doesn't crash or error on macOS, but I didn't check if it actually does what it's supposed to do. I'll just remove the usage of the script, and we can figure out X server stuff for macOS separately if we decide it's useful

**Review by tomtseng — COMMENTED** [review]
(leaving some comments to make this easier to review)

File: .circleci/config.yml [hunk 2]
```diff
@@ -42,25 +42,24 @@ commands:
...
-    description: "Check out and update Python dependencies."
+    description: "Check out and update Python dependencies on Linux."
     steps:
       - checkout
 
       # Download and cache dependencies
       - restore_cache:
           keys:
-            - v6-unix-dependencies-{{ checksum "setup.py" }}
+            - v6-linux-dependencies-{{ checksum "setup.py" }}-{{ checksum "ci/build_and_activate_venv.sh" }}
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply .circleci/config.yml:52]: agreed

File: .circleci/config.yml [hunk 3]
```diff
@@ -124,6 +160,45 @@ commands:
...
+  restore-pytest-cache:
+    description: "Restore .pytest_cache from CircleCI cache."
+    steps:
+      # The setup.py checksum in the cache name ensures that we retrain our expert
+      # policies, which are cached in the pytest cache, whenever setup.py changes. This
+      # is just a rough heuristic to decide when the experts should be retrained but
+      # still better than nothing.
+      - restore_cache:
+          keys:
+            - v7-pytest-cache-{{ arch }}-{{ checksum "setup.py" }}
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply .circleci/config.yml:172]: The cache should be compatible across architectures as it's just saving policies that are serialized in a standard format. But I think there is no harm to separating the architectures: it doesn't slow things down, as we run the tests in different archs in parallel so they'd all regenerate the cache upon validation and then it's just a race which one saves it first.

**Review by AdamGleave — APPROVED** [review]
LGTM

**Pull request merged**